### PR TITLE
Collect metrics for the `ic-canister`-derived canisters

### PR DESCRIPTION
**Pull request by awkure**

### CODE REVIEW

File: ic-canister/ic-canister-macros/src/derive.rs [hunk 1]
```diff
@@ -185,11 +248,48 @@ pub fn derive_canister(input: TokenStream) -> TokenStream {
...
+                let new_ts =
+                    if current_ts < last_ts + #interval {
+                        last_ts
+                    } else {
+                        current_ts - (current_ts % #interval)
+                    };
+                metrics
+                    .entry(new_ts)
+                    .and_modify(|e| { *e = new_metrics.clone() })
+                    .or_insert(new_metrics);
```
Comment (hagsteel) [inline ic-canister/ic-canister-macros/src/derive.rs:281]:
```suggestion
                metrics.insert(new_ts, new_metrics);
```

File: ic-canister/ic-canister/src/lib.rs [hunk 2]
```diff
@@ -154,6 +154,68 @@
 //!
 //! The API methods must be instance methods (taking `self` by reference).
 //!
+//! # Metrics
+//!
+//! Metrics for the canister can be collected by adding `#[metrics]` field to the canister struct.
```
Comment (hagsteel) [inline ic-canister/ic-canister/src/lib.rs:159]:
```suggestion
//! Metrics for the canister can be collected by adding a `#[metrics]` field to the canister struct.
```

File: ic-canister/ic-canister/src/lib.rs [hunk 3]
```diff
@@ -154,6 +154,68 @@
...
+//!     
+//!     #[metrics]
+//!     metrics: std::rc::Rc<RefCell<ic_canister::MetricsMap<Metrics>>>,
+//!
+//!     #[state]
+//!     state: Rc<RefCell<MyCanisterState>>,
+//! }
+//! ```
+//!
+//! Note that `Metrics` is wrapped under the [`MetricsMap`] struct. This is an implementation
```
Comment (hagsteel) [inline ic-canister/ic-canister/src/lib.rs:180]:
```suggestion
//! Note that `Metrics` is wrapped inside the [`MetricsMap`] struct. This is an implementation
```

File: ic-canister/ic-canister/src/lib.rs [hunk 4]
```diff
@@ -154,6 +154,68 @@
...
+//! detail used storing the metrics properly.
```
Comment (hagsteel) [inline ic-canister/ic-canister/src/lib.rs:181]:
```suggestion
//! detail used for storing the metrics properly.
```

File: ic-canister/ic-canister/src/lib.rs [hunk 5]
```diff
@@ -154,6 +154,68 @@
...
+//!
+//! Under the hood this adds two calls to the `MyCanister` that utilitzes `metrics` state: `collect_metrics()` and `get_metrics()`.
```
Comment (hagsteel) [inline ic-canister/ic-canister/src/lib.rs:183]:
```suggestion
//! Under the hood this adds two calls to the `MyCanister` struct that utilizes `metrics` state: `collect_metrics()` and `get_metrics()`.
```

File: ic-canister/ic-canister/src/lib.rs [hunk 6]
```diff
@@ -154,6 +154,68 @@
...
+//!
+//! With that in place we need to set up the logic for metrics calculation. This is done by setting
```
Comment (hagsteel) [inline ic-canister/ic-canister/src/lib.rs:185]:
```suggestion
//! With that in place we need to set up the logic for metrics calculation. This is done by setting the
```

File: ic-canister/ic-canister/src/lib.rs [hunk 7]
```diff
@@ -154,6 +154,68 @@
...
+//! `new_metric_snapshot` query method (that will ask you to set up if it's not yet present).
```
Comment (hagsteel) [inline ic-canister/ic-canister/src/lib.rs:186]:
```suggestion
//! `new_metric_snapshot` query method (see example below)
```

File: ic-canister/ic-canister/src/lib.rs [hunk 8]
```diff
@@ -154,6 +154,68 @@
...
+//!     }
+//! }
+//! ```
+//!
+//! This method is pure in a sense that we don't have to think about how we should update the
+//! `metrics` field itself. Only the way the new metric is collected.
+//!
+//! You can set up interval for metrics collection by specifiying it explicitly (by default it collects
+//! metrics every day) via `#[metrics(interval = interval_value)]` where `interval_value` could be either
+//! a number of minutes or an ident from the list `[min, hour, day, week, month]`.
```
Comment (hagsteel) [inline ic-canister/ic-canister/src/lib.rs:201]:
```suggestion
//! You can specify an interval for metrics collection by setting it explicitly (by default it collects
//! metrics every day) via `#[metrics(interval = interval_value)]` where `interval_value` is either
//! the number of minutes or an ident from the list `[min, hour, day, week, month]`.
```

File: ic-canister/ic-canister/src/lib.rs [hunk 9]
```diff
@@ -154,6 +154,68 @@
...
+//!     
+//!     #[metrics(interval = hour)]
+//!     metrics: std::rc::Rc<RefCell<ic_canister::MetricsMap<Metrics>>>,
+//!
+//!     #[state]
+//!     state: Rc<RefCell<MyCanisterState>>,
+//! }
+//! ```
+//!
+//! For an example you can refer to the tests in `canister_b` crate.
```
Comment (hagsteel) [inline ic-canister/ic-canister/src/lib.rs:217]:
```suggestion
//! For an example you can refer to the tests in the `canister_b` crate.
```

**Review by hagsteel — CHANGES_REQUESTED** [review]
I think perhaps the interval idents should really be:
* every-minute
* hourly
* daily
* weekly
* monthly
or something to that avail.
Let's discuss this in chat

File: ic-canister/ic-canister-macros/src/derive.rs [hunk 10]
```diff
@@ -28,6 +28,35 @@ impl Default for TraitNameAttr {
...
+            return Ok(Interval { mins: value });
+        }
+
+        let ident = input.parse::<syn::Ident>()?;
+        let interval = match ident.to_string().as_str() {
+            "min" => Interval { mins: 1 },
+            "hour" => Interval { mins: 60 },
+            "day" => Interval { mins: 60 * 24 },
+            "week" => Interval { mins: 60 * 24 * 7 },
+            "month" => Interval { mins: 60 * 24 * 30 },
```
Comment (hagsteel) [inline ic-canister/ic-canister-macros/src/derive.rs:51]: Also update anywhere in the docs / examples where this is used
```suggestion
            "hourly" => Interval { mins: 60 },
            "daily" => Interval { mins: 60 * 24 },
            "weekly" => Interval { mins: 60 * 24 * 7 },
            "monthly" => Interval { mins: 60 * 24 * 30 },
```

**Review by hagsteel — APPROVED** [review]
Looks good!

**Review by Maximkaaa — COMMENTED** [review]
I'd rather we go with the trait approach:
```
pub trait MetricsCanister : Canister {
  ...
}
```

**Comment by awkure** [comment]
Closing in favor https://github.com/infinity-swap/ic-helpers/pull/58

**Pull request closed without merging**